### ISSUE TITLE
(edge) "Root scope" doesn't work in components

### ISSUE THREAD

**Issue by dagnelies** [ractivejs/ractive#2432]
Hi,

here's an edge bug, with the fiddle as usual:

http://jsfiddle.net/xc2o8y08/45/

**Comment by martypdx** [comment]
The change from latest to edge was to no longer allowing a `~/foo` get to traverse up the component stack if it was not found at root of component. Arguably in edge it is now correct. 

It works in the "both" case because the unqualified reference `{{foo}}` creates a mapping locally at `foo` to the actual parent model value. Thus, the execution of `{{~/foo}}` succeeds.

But this is a rather subtle point of reference resolution, so open to other interpretations...

**Comment by dagnelies** [comment]
So... if I understand right ...in latest `~/` referenced the root of the ractive instance, and in edge the `~/` now references the root of the component ...right?

Therefore, you can still reference data in ractive's root, using anything except `~/` ...it's rather counter-intuitive. :dizzy_face:  

"Intuition wise" I'd have expected that this would happen only for isolated components.

**Comment by martypdx** [comment]
One possible addition would be to add `^/foo` to mean look anywhere above current context. It could then be combined to represent cases (which I find odd to be honest) like : `~/^/foo` meaning go to the root of this instance, then look up for first `foo`.

**Comment by evs-chris** [comment]
I think at that point you're better off with an alias, assuming all the refs are only in the template. It's a bit more explicit.

**Comment by guilhermeaiolfi** [comment]
> That could also open the door for the request to have an ractive option to only use restricted references (thus avoiding unnecessary resolution without having to use .foo). {{name}} means ```{{.name}}``` and ```{{^/name}}``` means look up the stack. 

I've suggested that in another issue. That would be the sane thing to do IMO. And use ```restrictedRefs``` as ```true``` by default as well. :+1:

**Comment by martypdx** [comment]
> I think at that point you're better off with an alias, assuming all the refs are only in the template. It's a bit more explicit.

@evs-chris in that example, sure you could do that. (side note: would be nice to add alias block that doesn't change context, will open separate issue for that...)

But doesn't change the issue around `~/` being non-restricted. I had to always put default data in the component simply to prevent the resolution from walking the context chain.

**Comment by dagnelies** [comment]
> No. In latest ~/ also referenced the root of the component instance, but if nothing was found, it continued up the context stack into the parent 

Ahaaaa ...now I get it. Thanks for enlightening me.

Hmmm ...I guess both are valid ways of doing things (considering `~/` to reference the instance root or the component root). It's really not obvious to determine what is actually the most intuitive. Most importantly, I'll leave this ticket open until I've submitted a PR to add this information to the docs, which is rather important.

As for the whole adding syntax stuff, I'm personally against. There's already so much that people will get overwhelmed if we add more. I'm personally OK with the situation as it is now, you just have to be aware of it.

**Comment by dagnelies** [comment]
Here it is: #249 It's just a single line but better than nothing.

**Comment by dagnelies** [comment]
Hi,

...back to the topic ...after sleeping over it, I think the most intuitive is still the current `latest` behavior.

Let me explain. `~/` is a bit ambiguous, because it could be interpreted intuitively both as the root of the component or the root of the instance. From that point of view, it kinds of makes sense to try first a lookup at the component root, and if there is none, try at the ractive root. For most usecases, it would probably resolve to ractive's root, but offer the possibility to override it in the component. I think that would be the preferred way of handling things, like in 0.7

**Comment by evs-chris** [comment]
__TL;DR avoid ambiguity and `{{(╯°□°)╯︵ ┻━┻/foo}}` :smile:__

I have to disagree because the `~/` should _remove_ any ambiguity. It should either be the root of the local instance or the root of the root instance. The latter would make non-isolated components unable to reach non-inited data at their root. It makes more sense to me to introduce some sort of context-hopping prefix than to remove precision from an existing (_should be_) precise prefix. I have certainly found scenarios where being able to express "I want to jump up to the context in the section containing this one and access the foo member of its keypath parent". That was one of the drivers for aliases - accessing context that was inaccessible without having to make sure ambiguity wouldn't land in the wrong place:

```
{{#some.key.path}}
  {{#~/other}} I also have a foo key
    how do I say some.key.foo?
    granted I could be explicit here, but if this was a partial, I couldn't
  {{/}}
{{/}}
```

```
{{#some.key.path}}
  {{#with ../ as context}}
    {{#~/other}} I also have a foo key
      hey, it's {{context.foo}}
    {{/}}
  {{/with}}
{{/}}
```

the alternate as suggested by @martypdx, and I _would_ avoid `^` because it also means `unless`, but nothing else looked right
```
{{#some.key.path}}
  {{#~/other}} I also have a foo key
    try {{^/../foo}} which is absolutely explicit
  {{/}}
{{/}}
```

Aside from being _one more_ thing to look at, that's really handy, especially when stacked: `^/^/^/../foo` = pop context thrice, walk up the keypath one level, and give me foo. Combining context pops and root refs _would_ be nice for everyone dealing with larger projects and collaboration that want to be able to remove any resolution surprises that result from ambiguity. Going crazy with covering all of the bases, we have ops for popping keypath levels and accessing the root of the current keypath (`..` and `~`). Popping contexts (the proposed `^`) would need a root context analog for `~` (`~^`?), and component levels (`^^` and `~^^`?) should probably be addressed as well for non-isolated components. I don't see hopping components as being terribly crucial since it would go against best-practices for building reusable components.

**Comment by JonDum** [comment]
> Popping contexts (the proposed ^) would need a root context analog for ~ (~^?), and component levels (^^ and ~^^?)

You're just making smiley faces at this point, Chris :P

**Comment by fskreuz** [comment]
Not really a fan of traversing the context, but can't these be modeled like paths? Like:

- `foo/bar` - Starting from the current context, get `foo.bar`
- `../foo/bar` - Starting from the current context, climb one up, then `foo.bar`
- `~/foo/bar` - (leading tilde) Starting from the top of the component, get `foo.bar`
- `/foo/bar` - (leading slash) Starting from the very top level component, get `foo.bar`
- `foo/bar.baz../bam` - (mixed mode) Get `foo.bar.bam`.

Dot notation makes sense in vanilla JS terms. But when traversing contexts (and where dots mean something else), slashes make sense. Does this make sense?

**Comment by JonDum** [comment]
That's been brought up before. The problem is that the root slash looks like a closing tag:

`{{/foo}}`

**Comment by fskreuz** [comment]
I believe the names in closing tags are optional in most cases (except for some helpers). Can we drop them altogether so all closing tags will be like `{{/}}` (unless `{{/}}` should also print out the root object).

**Comment by dagnelies** [comment]
Well, you're all right. I agree with @evs-chris that my "ambiguity fallback" proposal was not a good idea. `~/` should resolve to one and only one place.

What about changing `~/` to simply mean the uppermost level then. The true "root". This would be the ractive instance root for non-isolated components, and the component itself for isolated ones. Still pretty intuitive, right?
